### PR TITLE
fix(provider): Manage context correctly, so the provider can shutdown

### PR DESCRIPTION
**Pull request by hydrogen18**
I couldn't get my provider to shutdown when I had a bunch of open deployments that the provider was trying to bid on.

I ended up making the following changes to get this to work

1. When the bidengine service starts, the service queries for existing orders. But it does not query for the bids
2. When the bidengine order starts, it will query for an existing bid _if an order already existed when the provider was started_. So this query-ing happens in parallel rather than serially in a loop
3. Don't attempt to store the bid in the bidengine order handler, instead just keep a flag indicating if it has bid already. This avoids some issues where the bidengine order handler thought a bid existed with an ID of `0/0/0` which is obviously invalid
4. Don't ignore errors when querying for the existing bid _unless_ the error message says "bid not found"
5. When the bid engine order handler stops and the bid _is not won_ always cancel the bid. This previously did not happen
6. I chased down places we were just using `context.Background()` and then refactored them into a managed context within the same file that is cancelled when the goroutine gets a shutdown signal.
7. Increased the test coverage around the bid engine order handler.

### CODE REVIEW

File: provider/cluster/manager.go [hunk 1]
```diff
@@ -94,6 +94,7 @@ func (dm *deploymentManager) teardown() error {
 
 func (dm *deploymentManager) run() {
 	defer dm.lc.ShutdownCompleted()
+	ctx, cancel := context.WithCancel(context.Background())
```
Comment (boz) [inline provider/cluster/manager.go:97]: The thing with this one is that it's not clear that we should actually cancel operations when things start to shut down.

For instance if it's in the middle of closing a lease and we bounce the provider, we don't really want to cancel the api call to delete the namespace.

File: provider/bidengine/order.go [hunk 2]
```diff
@@ -74,35 +76,75 @@ func newOrderInternal(svc *service, oid mtypes.OrderID, bid *mtypes.Bid, pricing
 	return order, nil
 }
 
-func (o *order) run() {
+var matchBidNotFound = regexp.MustCompile("^.+bid not found.+$")
+
+func (o *order) run(checkForExistingBid bool) {
 	defer o.lc.ShutdownCompleted()
+	// TODO -get context from parent
```
Comment (boz) [inline provider/bidengine/order.go:83]: It's okay - this gets cancelled when we're supposed to shut down.

File: provider/bidengine/order.go [hunk 3]
```diff
@@ -74,35 +76,75 @@ func newOrderInternal(svc *service, oid mtypes.OrderID, bid *mtypes.Bid, pricing
...
 
+		case queryBid := <-queryBidCh:
+			err := queryBid.Error()
+			bidFound := true
+			if err != nil {
+				if matchBidNotFound.MatchString(err.Error()) {
+					bidFound = false
+				} else {
+					o.session.Log().Error("could not get existing bid", "err", err, "errtype", fmt.Sprintf("%T", err))
+					break
```
Comment (boz) [inline provider/bidengine/order.go:136]: `break loop`?  can we recover from this?

File: provider/bidengine/order.go [hunk 4]
```diff
@@ -74,35 +76,75 @@ func newOrderInternal(svc *service, oid mtypes.OrderID, bid *mtypes.Bid, pricing
...
+				}
+			}
+
+			if bidFound {
+				o.bidPlaced = true
+				_ = queryBid.Value().(*mtypes.QueryBidResponse) // TODO do I need me?
```
Comment (boz) [inline provider/bidengine/order.go:142]: no.

File: provider/bidengine/order.go [hunk 5]
```diff
@@ -228,6 +267,8 @@ loop:
 			}
 
 			// Fulfillment placed.
+			o.bidPlaced = true
+			result.Value()
```
Comment (boz) [inline provider/bidengine/order.go:271]: I don't think it's needed

File: provider/bidengine/order_test.go [hunk 6]
```diff
@@ -141,13 +142,35 @@ func makeOrderForTest(t *testing.T, bid *mtypes.Bid, pricing BidPricingStrategy)
...
+			},
+		}
+		scaffold.queryClient.On("Bid", mock.Anything, queryBidRequest).Return(response, nil)
+	}
+	/**
 	if bid != nil {
 		bid.BidID = mtypes.MakeBidID(scaffold.orderID, scaffold.testAddr)
 		bid.Price = testutil.AkashCoin(t, 1)
-	}
+	}**/
```
Comment (boz) [inline provider/bidengine/order_test.go:170]: :scissors:

File: provider/cluster/monitor.go [hunk 7]
```diff
@@ -176,10 +177,10 @@ func (m *deploymentMonitor) doCheck() (bool, error) {
 	return badsvc == 0, nil
 }
 
-func (m *deploymentMonitor) runCloseLease() <-chan runner.Result {
+func (m *deploymentMonitor) runCloseLease(ctx context.Context) <-chan runner.Result {
 	return runner.Do(func() runner.Result {
 		// TODO: retry, timeout
-		err := m.session.Client().Tx().Broadcast(context.Background(), &mtypes.MsgCloseBid{
+		err := m.session.Client().Tx().Broadcast(ctx, &mtypes.MsgCloseBid{
```
Comment (boz) [inline provider/cluster/monitor.go:183]: we might have to do a reaper task that closes leases in addition to the one we've talked about that cleans up namespaces.

**Review by boz — COMMENTED** [review]
Looks great, some comment inline.

File: provider/cluster/manager.go (hunk 1 above)
Comment (hydrogen18) [reply provider/cluster/manager.go:97]: Yeah, I agree we'd rather wait on that to finish (even if it takes a while). I'll go back over this and try and make sure I'm not messing anything up like that.

File: provider/bidengine/order.go (hunk 3 above)
Comment (hydrogen18) [reply provider/bidengine/order.go:136]: I was kind of wondering that as well. There are two possible outcomes when we get an error here 

1. Try and recover - provider might send another bid created event, even if one already exists. What happens in that case?
2. Bail completely - provider misses out on bidding on this deployment?

I'm not sure which one makes the most sense.

Option 3 - retry (with a delay) forever until we get past the error?

File: provider/bidengine/order.go (hunk 4 above)
Comment (hydrogen18) [reply provider/bidengine/order.go:142]: Yeah I was leaning to that, I'll kill this line.

File: provider/bidengine/order.go (hunk 5 above)
Comment (hydrogen18) [reply provider/bidengine/order.go:271]: Yeah the `result.Value()` isn't doing anything anymore. I'll kill it

**Comment by hydrogen18** [comment]
@boz I think I've addressed all your comments now

**Review by boz — APPROVED** [review]
:pray: :rocket:

**Pull request merged**